content/v2.0/reference/internals/storage-engine.md

```diff
@@ -0,0 +1,106 @@
+
+Major topics include:
+
+* [Write Ahead Log (WAL)](#write-ahead-log-wal)
+* [Cache](#cache)
+* [Time-Structed Merge Tree (TSM)](#time-structured-merge-tree-tsm)
+* [Time Series Index (TSI)](#time-series-index-tsi)
+
+## Writing data: from API to disk
+
+The storage engine handles data from the point an API write request is received through writing data to the physical disk.
+Data is written to InfluxDB using [line protocol](/v2.0/reference/line-protocol/) sent via HTTP POST request to the `/write` endpoint.
+Batches of [points](/v2.0/reference/glossary/#point) are sent to InfluxDB, compressed, and written to a WAL for immediate durability.
+Points are also written to an in-memory cache and become immediately queryable.
+The cache is periodically written to disk in the form of [TSM](#time-structured-merge-tree-tsm) files.
+As TSM files accumulate, they are combined and compacted into higher level TSM files.
+
+Points can be sent individually; however, for efficiency, most applications send points in batches.
+Points in a POST body can be from an arbitrary number of series, measurements, and tag sets.
+Points in a batch do not have to be from the same measurement or tagset.
+
+## Write Ahead Log (WAL)
+
+The **Write Ahead Log** (WAL) ensures durability by retaining data when the storage engine restarts.
+It ensures that written data does not disappear in an unexpected failure.
+When a client sends a write request, the following steps occur:
+
+1. Append write request to the end of the WAL file.
+2. Write data to disk using `fsync()`.
+3. Update the in-memory cache.
+4. Return success to caller.
+
+`fsync()` takes the file and pushes pending writes all the way to the disk.
+As a system call, `fsync()` has a kernel context switch which is computationally expensive, but guarantees that data is safe on disk.
+
+{{% note%}}
+Once you receive a response to a write request, your data is on disk!
+{{% /note %}}
+
+When the storage engine restarts, the WAL file is read back into the in-memory database.
+InfluxDB then answers requests to the `/read` endpoint.
+
+## Cache
+
+The **cache** is an in-memory copy of data points currently stored in the WAL.
+Points are organized by key, which is the measurement, tag set, and unique field.
+Each field is stored in its own time-ordered range.
+Data is not compressed in the cache.
+The cache is recreated on restart by re-reading the WAL files on disk back into memory.
+The cache is queried at runtime and merged with the data stored in TSM files.
+When the storage engine restarts, WAL files are re-read into the in-memory cache.
+
+Queries to the storage engine will merge data from the cache with data from the TSM files.
+Queries execute on a copy of the data that is made from the cache at query processing time.
+This way writes that come in while a query is running do not affect the result.
+Deletes sent to the cache will clear out the given key or the specific time range for the given key.
```

Comment:
Deletes sent to the cache clear the specified key or time range for a specified key.